### PR TITLE
Allow tooling API `BuildAction` instances to stream results to the client application

### PR DESCRIPTION
**Pull request by adammurdoch**
### Context

This PR adds 2 public APIs:
- `BuildController.sendIntermediate()` is used by `BuildAction` implementations to queue an object to send back to the client.
- `BuildActionExecuter.setIntermediateModelListener()` is used by the client to provide a listener that will receive these objects.

This PR intentionally does no attempt to make the APIs work well with older versions of Gradle. The methods currently fail with an exception or silently do nothing when used with older versions.

### Contributor Checklist
- [ ] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [ ] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [ ] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [ ] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [ ] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [ ] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

File: platforms/core-runtime/launcher/src/main/java/org/gradle/tooling/internal/provider/IntermediateModelListenerAdapter.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    public IntermediateModelListenerAdapter(ProviderOperationParameters providerParameters, PayloadSerializer payloadSerializer, BuildEventConsumer delegate) {
+        this.providerParameters = providerParameters;
+        this.payloadSerializer = payloadSerializer;
+        this.delegate = delegate;
+    }
+
+    @Override
+    public void dispatch(Object message) {
+        if (message instanceof IntermediateModel) {
+            if (failure != null) {
```
Comment (bamboo) [inline platforms/core-runtime/launcher/src/main/java/org/gradle/tooling/internal/provider/IntermediateModelListenerAdapter.java:43]: 💅 lines 43-52 could be extracted to a separate method so the statements in this methods are at the same level of abstraction

File: platforms/ide/tooling-api/src/main/java/org/gradle/tooling/BuildActionExecuter.java [hunk 2]
```diff
@@ -69,6 +70,15 @@ interface Builder {
         BuildActionExecuter<Void> build();
     }
 
+    /**
+     * Sets the listener to use to handle intermediate results of type {@code S} sent from the action via {@link BuildController#sendIntermediate(Object)}.
```
[the remark this answers is not in the document]
Comment (adammurdoch) [reply platforms/ide/tooling-api/src/main/java/org/gradle/tooling/BuildActionExecuter.java]: This is fix.

**Comment by adammurdoch** [comment]
@bot-gradle test and merge

File: platforms/core-runtime/launcher/src/main/java/org/gradle/tooling/internal/provider/IntermediateModelListenerAdapter.java (hunk 1 above)
Comment (adammurdoch) [reply platforms/core-runtime/launcher/src/main/java/org/gradle/tooling/internal/provider/IntermediateModelListenerAdapter.java:43]: This is done in https://github.com/gradle/gradle/pull/27461

**Pull request merged**